AdyenActions/UI/View Controllers/QR Code/QRCodeViewModel.swift

```diff
@@ -4,81 +4,125 @@
 // This file is open source and available under the MIT license. See the LICENSE file for more info.
 //
 
-@_spi(AdyenInternal) import Adyen
 import UIKit
+@_spi(AdyenInternal) import Adyen
 
-extension QRCodeView {
-
-    internal class Model {
-
-        internal enum ActionButton {
-            case copyCode
-            case saveAsImage
-        }
+internal protocol QRCodeViewModelProtocol {
+    var flowType: QRCodeFlowType { get }
+    var instructionText: String { get }
+    var amountText: String? { get }
+    var actionButtonTitle: String { get }
+    var onCopyButtonTitle: String { get }
+    var qrCodeData: String { get }
+    var expiration: AdyenObservable<String?> { get }
+    var copyInProgress: AdyenObservable<Bool> { get }
+    var observedProgress: Progress? { get }
+    
+    func loadLogoImage(completion: @escaping (UIImage?) -> Void)
+    func performAction(qrCodeImage: UIImage?, from: UIView)
+}
 
-        internal let action: QRCodeAction
+internal class QRCodeViewModel: QRCodeViewModelProtocol, Localizable {
+    
+    private enum Constants {
+        static let copyAnimationDuration: TimeInterval = 2
+    }
+    
+    // MARK: - Properties
         
-        internal var actionButtonType: ActionButton {
+    private let action: QRCodeAction
+    private let logoUrl: URL
+    private let imageLoader: ImageLoading
+    private let onSaveQRCode: (_ image: UIImage?, _ sourceView: UIView) -> Void
+    private let onCopyCode: (_ code: String) -> Void
+    internal let observedProgress: Progress?
+    internal let expiration: AdyenObservable<String?>
+    internal var copyInProgress: AdyenObservable<Bool> = .init(false)
+    internal var localizationParameters: LocalizationParameters?
+    
+    internal let instructionText: String
+    internal let amountText: String?
+    internal let flowType: QRCodeFlowType
+    
+    // MARK: - Initializers
+    
+    internal init(
+        action: QRCodeAction,
+        instructionText: String,
+        payment: Payment?,
+        logoUrl: URL,
+        observedProgress: Progress?,
+        expiration: AdyenObservable<String?>,
+        imageLoader: ImageLoading = ImageLoaderProvider.imageLoader(),
+        localizationParameters: LocalizationParameters?,
+        onSaveQRCode: @escaping (_ image: UIImage?, _ sourceView: UIView) -> Void,
+        onCopyCode: @escaping (_ code: String) -> Void
+    ) {
+        self.action = action
+        self.instructionText = instructionText
+        self.amountText = payment?.amount.formatted
+        self.logoUrl = logoUrl
+        self.observedProgress = observedProgress
+        self.expiration = expiration
+        self.imageLoader = imageLoader
+        self.localizationParameters = localizationParameters
+        self.onSaveQRCode = onSaveQRCode
+        self.onCopyCode = onCopyCode
+        
+        self.flowType = {
```

Comment:
should we move this logic to check `flowType` to separate private method for clarity?

Reply:
yes, computed property is good to have :)